File: design_patterns/002_observer/solutions/typescript/weatherStation/src/DisplayCurrentConditions.ts
import { IDisplay } from './IDisplay';
import { IObserver } from './IObserver';
import { WeatherData, WeatherInfo } from './WeatherData';

export class DisplayCurrentConditions implements IObserver, IDisplay {
  private _weather_data: WeatherData | null = null;
  private _temperature: number | null = null;
  private _humidity: number | null = null;

  constructor(weather_data: WeatherData) {
    this._weather_data = weather_data;
    weather_data.subscribe(this);
  }

  update(updated_data: { [key: string]: any }): void {
    if (updated_data.hasOwnProperty(WeatherInfo.TEMPERATURE)) {
      this._temperature = updated_data[WeatherInfo.TEMPERATURE];
    }
    if (updated_data.hasOwnProperty(WeatherInfo.HUMIDITY)) {
      this._humidity = updated_data[WeatherInfo.HUMIDITY];
    }
    this.display();
  }

  display(): void {
    if (this._temperature !== null && this._humidity !== null) {
      console.log(`Current conditions: ${this._temperature} F degrees and ${this._humidity}% humidity`);
    }
  }
}

File: design_patterns/002_observer/solutions/python/weatherStation/src/DisplayCurrentConditions.py
from IDisplay import IDisplay
from IObserver import IObserver
from IPublisher import IPublisher
from WeatherData import WeatherData, WeatherInfo


class DisplayCurrentConditions(IObserver, IDisplay):
    _weather_data:WeatherData = None
    _temperature:float = None
    _humidity:float = None


    def __init__(self, weather_data:WeatherData) -> None:
        self._weather_data = weather_data
        weather_data.subscribe(self)


    def update(self, updated_data: dict) -> None:
        if temp := updated_data.get(WeatherInfo.TEMPERATURE):
            self._temperature = temp
        if humidity := updated_data.get(WeatherInfo.HUMIDITY):
            self._humidity = humidity
        self.display()


    def display(self) -> None:
        print(f'Current conditions: {self._temperature} F degrees and {self._humidity}% humidity')


if __name__ == '__main__':
    raise Exception(f'This module should not be executed directly. Only for imports')
